Add new regular-chatbot concept exercise to the JavaScript Track (#1653)
exercism/javascript: concepts/regular-expressions/about.md

# About

## Regular Expressions in JavaScript

A [Regular Expression][mdn-regular-expressions] (or Regex) is a sequence of characters that we can use to target and manipulate certain elements in strings. Hence, we can:

- **Search** for a text in a string
- **Replace** substrings in a string
- **Extract** information from a string

> 💡 JavaScript's regex flavor is part of the ECMA-262 standard for the language. This means that you don't have to worry about browser-specific compatibility.

> 

Reviewer: I would recommend to add a sentence here like:
"This concept is focused on using regular expressions in JavaScript. If you want to learn about how to write regular expressions, [Eloquent JavaScript](https://eloquentjavascript.net/09_regexp.html) might be a good resource."

(The link needs to be converted to our link format we use.)

Author: Thanks for the recommendation! I also think it is a great resource that will complement perfectly the exercise. 

I'll add that to the Introduction.

## How to create Regular Expressions

1. **Regular Expression Literal**:

```javascript
const regex = /[a-z]/;
```

2. Constructor **RegExp**:

```javascript
const regex = new RegExp('[a-z]');
```

## 🏴‍☠️ Flags

Some of the widely used are:

- `/g` - Global Search
- `/i` - Case Insensitive
- `/m` - Multiline Search

Here is a simple example:

```javascript
const re = /home/gi;
const str = 'Home, sweet home.';

str.match(re);

// => ["Home", "home"]
```

## Commonly used functions

When regular expressions are combined with current build-in functions in JavaScript, we have some powerful ways of manipulating and getting data from strings.

> The ``test()`` method is a great way of searching and knowing if a target value exists within a given string. In this way, it returns a boolean value, `true` or `false`. 

Author: For some reason, I thought it was with two back-ticks, maybe due to the markdown in the mentor suggestions. 

Thanks for noticing it. I'll stick with single back-ticks.

The ``match()`` method extracts the regular expression match from a given string. Returning an array with information and matches.

> `When to use one or the other?`  Use the ``test()`` when you want a fast way of checking a value within a string, use ``match()`` if you need to use that value or want more information about the match. 

Author: Got that. I actually thought I could also emphasize with single back-ticks. I'll change that phrase into italic form.

# About

> ## 1. Regular Expressions in JavaScript

Reviewer: We don't user numbers in headers anywhere else so it would be better to remove them. That also makes it easier for contributors to add/change things later.

A [Regular Expression][mdn-regular-expressions] (or Regex) is a sequence of characters that we can use to target and manipulate certain elements in strings. Hence, we can:

- **Search** for a text in a string
- **Replace** substrings in a string
- **Extract** information from a string

> 💡 JavaScript's regex flavor is part of the ECMA-262 standard for the language. This means that you don't have to worry about browser-specific compatibility.

## 2. How to create Regular Expressions

> 

Reviewer: An introduction sentence about that there are 2 ways to create regex and when to use which would be nice. For performance concerns you can say "see below" but might be good to mention that the constructor can be used when building regex dynamically from strings.

Reviewer: I am a bit confused here. This was marked resolved but there is no change visible.

Author: @junedev Oh I'm sorry if I forgot to change that! Please I truly appreciate every comment and suggestion you are writing, I tried to change one by one on my local files. The problem is that I am not really familiar with these suggestions, resolving and implementing workflow. 

Could you tell me what is the most standard procedure?

I completely agree with every comment 💪🏽If I didn't implement the suggestions is just because I didn't see it or I didn't put enough effort into it. Sorry about that. Next time I'll try to be more careful. 

As soon as I have time, I'll work on this introductory sentence and the mention for the constructor. 

All the love.

1. **Regular Expression Literal**:

```javascript
const regex = /[a-z]/;
```

2. Constructor **RegExp**:

```javascript
const regex = new RegExp('[a-z]');
```

## 3. 🏴‍☠️ Flags

Some of the widely used are:

- `/g` - Global Search
- `/i` - Case Insensitive
- `/m` - Multiline Search

Here is a simple example:

```javascript
const re = /home/gi;
const str = 'Home, sweet home.';

> const myArray = str.match(re);
> console.log(myArray);
> 
> => // ["Home", "home"]

Reviewer: We often use this REPL style notation to make examples more concise:

```suggestion
str.match(re);
// => ["Home", "home"]
```

So just an expression and then `// =>` in the next line followed by the result of the expression.

It is also ok to keep your console log version but `=> //` needs to be changed to `// =>` in any case.

```

```javascript
let regex = /[a-z]/gi; // literal notation
let regex = new RegExp('[a-z]', 'gi'); // constructor with string pattern as first argument

> let regex = new RegExp(/[a-z]/, 'gi'); // constructor with regular expression literal as first argument (Starting with ECMAScript 6)

Reviewer: The assignments don't add value here, they could be removed.

```

> ## 4. Most common Functions

Reviewer: "Commonly used functions" might be better

When regular expressions are combined with current build-in functions in JavaScript, we have some powerful ways of manipulating and getting data from strings.

These are some of the most common functions used alongside regex.

### Test

> 

Reviewer: Would be great if you could add one sentence here when to use test. I often see people using match when they only need `test` so I think it would be valuable to be explicit about this.

```javascript
const str = 'It is difficult to test if you have a virus';
const result = /virus$/.test(str);

console.log(result); // true
```

### Match

> In this way, we are able both to **search** and to **extract** information from any string. For example:

Reviewer: It is not clear what "In this way" refers to, maybe better:
```suggestion
Using `match`, we are able both to **search** and to **extract** information from any string.
```

"For example" can be removed as included in the suggestion above. It is not necessary to introduce each example block like this.

```javascript
const funnyQuote =
  'If you see someone crying, ask if it is because of their haircut.';
const regex1 = /someone/;
const regex2 = /happy/;

> funnyQuote.match(regex1); // ["someone", index: 3, input: "If you see someone crying, ask if it is because of their haircut.", groups: undefined]

Reviewer: I would recommend to put the output in a new line below the expression (prefixed with `// =>` as explained above). Applies for this whole section.

funnyQuote.match(regex2); // null
```

> When the Global Search flag `/g` is present, instead of getting the only match alongside useful information such as the index or input, the method returns all of the occurances displayed in the array:

Reviewer: ```suggestion
When the Global Search flag `/g` is present, instead of getting the only match alongside useful information such as the index or input, the method returns all of the occurrences present in the array:
```

```

### Replace

The [replace()][mdn-regex-replace] method in JavaScript allows us to search for a value within a given string, and replacing it with a desired new value.

```javascript
string.replace(searchValue, newValue);
```

The pattern for searching the substitution can be a single string, or a regular expression.

```javascript
let string = 'I like dogs!';
let result = string.replace('dogs', 'cats');

let string = 'I would love to travel to Japan';
let result = string.replace(/Japan/g, 'Hawaii');
```

Moreover, we can apply a function on the replacement position in order to make additional changes to each value.

```javascript
let text = 'Say hello to the chatbot.';
let result = text.replace(/chatbot|hello/gi, function (word) {
  return word.toUpperCase();
});
// -> "Say HELLO to the CHATBOT"
```

### Split

The [split()][mdn-regex-split] method in JavaScript represents a different way of using and manipulating strings with regular expressions.

```javascript
const str = 'hello,user.how are.you';

const result = str.split(/[,.\s]/);

console.log(result);
//--> ['hello', 'user', 'how', 'are', 'you']
```

## 5. Performance with Regex

> Regarding performance, both of them create a RegExp object. The main difference is how often the regex is compiled:

Reviewer: The reference for "both of them" is not clear here. To fix this and since you cover more than performance here, maybe it would be better to make this whole section part of "creating regex" above.

- With `Regular Expression Literal`: one time during initial code parsing and compiling
- With `RegExp()` syntax: Every time new Object gets created.

> According to [this](https://stackoverflow.com/a/32523333)answer from Alexander Abakumov in StackOverflow, using literal syntax may be a better option not only because of performance, but also for simplicity and readability.

Reviewer: If this is not some famous person, I would recommend to just list the points you want to make and then provide the link as "for more details see this Stackoverflow discussion".

1. It is shorter and doesn’t force you to think in terms of class-like constructors.
2. When using the `RegExp()` constructor, you also need to escape quotes and double-escape backslashes. It makes regular expressions that are hard to read and understand by their nature even more harder.

> Nevertheless, for this particular challenge, we are practising with both methods. Hence you are able to decide which one to apply depending on the occasion.

Reviewer: This part does not belong on the concept, only in the exercise introduction.

> ## 6. Useful Resources

Reviewer: This headline can be removed. The list below will not be rendered. Instead, the website will render the content of the links.json file with a "Read more" headline.
